### DISCUSSION TITLE
NPM workspaces install non symlink

### DISCUSSION THREAD

**Discussion by joezappie** [npm/feedback#774] [General]
When using npm workspaces, is it possible to install a workspace as a regular dependency instead of a symlink? I see `npm ci` has the option `install-links` but it says:

> When set file: protocol dependencies that exist outside of the project root will be packed and installed as regular dependencies instead of creating a symlink. This option has no effect on workspaces.

The reason I'd like this is to flatten my final docker image when building an app in a monorepo. Right now seems like the solution is to either build and manually copy each workspace into node_modules (not a fan of that) or leave the final images folder structure to start at the monorepo's root and include each workspace needed. I'd rather have my final image be just my source code and a single node_module folder with all the dependencies installed as if they were installed from a registry.

Example:

Instead of docker image file structure being:
```
monorepo
  - node_modules
    - @monorepo/common-> /common/lib1
  - apps
    - app1
      - src
  - common
    - lib1
```

I'd like it to end up being:
```
app1
  - node_modules
    - @monorepo/common
  - src
```

With a staged docker build and no symlinks, it can just COPY the entire node_module folder and discard the monorepo format.

Is there a reason install-links has no effect for workspaces?

**Comment by kaihendry** [comment]
Just thinking aloud, why do you need the node_modules, shouldn't you on worry about shipping some artefact, e.g. a `bundle.js` file?

**Comment by joezappie** [comment]
This was for shipping an internal (not a library or public image) docker container for one of my server applications, where I don't use a bundler so it needs to ship node_modules. I think its a pretty common thing to have an `npm install` script inside your docker container, but I'm also fairly new to Docker so I very well could be wrong. I've just seen pretty much ever example with an `npm install` step, where it gets copied into the final image.

This was more of a nice to have request. It works as it stands, theres just several drawbacks currently:

Method 1: Copy entire monorepo
1. Need to manually remove irrelevant workspaces
2. Final docker file structure needs to have the entire monorepo's folder structure instead of just your workspace code.

Method 2: Copy only the workspace for the project
1. Need to manually copy any workspaces its dependent on as npm only uses a symlink for them.
2. If you add another workspace as a dependency, need to remember to add it to the docker build file

Been a while since I wrote this, but I was mainly curious _why_ `install-links` doesn't work for workspaces for 'npm ci`. Whether its an entirely separate project that was `npm linked` or a local workspaces that's being `linked` to another workspace, I feel they should work the same when talking about a continuous integration environment.

**Comment by Justin1002** [comment]
Hey I also have this issue with workspaces, this primarily causes issues with your jest tests being unable to resolve the node_modules unless you build the specific workspace package. 

We don't want to have to build all packages in our workspace just to accomplish testing in the source.

**Reply by Justin1002** [reply]
Also I think a big issue is related to package dependency versions not being respected by the symbolic link.

IE) I have package A installed in my repository that depends on a workspace package B at version `1.0.0`. If package B is at version `2.0.0` IE, contains breaking changes, it will still attempt to symbolically link this package.

Let me know if i'm wrong about this!
